indexing-service/src/test/java/io/druid/indexing/overlord/TaskLockboxTest.java

```diff
@@ -207,4 +258,206 @@ public void testSyncFromStorage() throws EntryExistsException
+    }
+
+    final Task highPriorityTask = NoopTask.create(100);
+    lockbox.add(highPriorityTask);
+    taskStorage.insert(highPriorityTask, TaskStatus.running(highPriorityTask.getId()));
+    final TaskLock lock = lockbox.tryLock(TaskLockType.EXCLUSIVE, highPriorityTask, interval).getTaskLock();
+    Assert.assertNotNull(lock);
+
+    Assert.assertTrue(
+        lockbox.doInCriticalSection(
+            highPriorityTask,
+            Collections.singletonList(interval),
+            CriticalAction.<Boolean>builder().onValidLocks(() -> true).onInvalidLocks(() -> false).build()
+        )
+    );
+  }
+
+  @Test
+  public void testDoInCriticalSectionWithRevokedLock() throws Exception
+  {
+    final Interval interval = Intervals.of("2017-01-01/2017-01-02");
+    final Task lowPriorityTask = NoopTask.create("task1", 0);
+    final Task highPriorityTask = NoopTask.create("task2", 10);
+    lockbox.add(lowPriorityTask);
+    lockbox.add(highPriorityTask);
+    taskStorage.insert(lowPriorityTask, TaskStatus.running(lowPriorityTask.getId()));
+    taskStorage.insert(highPriorityTask, TaskStatus.running(highPriorityTask.getId()));
+
+    final TaskLock lowPriorityLock = lockbox.tryLock(TaskLockType.EXCLUSIVE, lowPriorityTask, interval).getTaskLock();
+    Assert.assertNotNull(lowPriorityLock);
+    Assert.assertTrue(lockbox.tryLock(TaskLockType.EXCLUSIVE, highPriorityTask, interval).isOk());
+    Assert.assertTrue(Iterables.getOnlyElement(lockbox.findLocksForTask(lowPriorityTask)).isRevoked());
+
+    Assert.assertFalse(
+        lockbox.doInCriticalSection(
+            lowPriorityTask,
+            Collections.singletonList(interval),
+            CriticalAction.<Boolean>builder().onValidLocks(() -> true).onInvalidLocks(() -> false).build()
+        )
+    );
+  }
+
+  @Test(timeout = 5000L)
+  public void testAcquireLockAfterRevoked() throws EntryExistsException, InterruptedException
+  {
+    final Interval interval = Intervals.of("2017-01-01/2017-01-02");
+    final Task lowPriorityTask = NoopTask.create("task1", 0);
+    final Task highPriorityTask = NoopTask.create("task2", 10);
+    lockbox.add(lowPriorityTask);
+    lockbox.add(highPriorityTask);
+    taskStorage.insert(lowPriorityTask, TaskStatus.running(lowPriorityTask.getId()));
+    taskStorage.insert(highPriorityTask, TaskStatus.running(highPriorityTask.getId()));
+
+    final TaskLock lowPriorityLock = lockbox.lock(TaskLockType.EXCLUSIVE, lowPriorityTask, interval).getTaskLock();
+    Assert.assertNotNull(lowPriorityLock);
+    Assert.assertTrue(lockbox.tryLock(TaskLockType.EXCLUSIVE, highPriorityTask, interval).isOk());
+    Assert.assertTrue(Iterables.getOnlyElement(lockbox.findLocksForTask(lowPriorityTask)).isRevoked());
+
+    lockbox.unlock(highPriorityTask, interval);
+
+    // Acquire again
+    final LockResult lockResult = lockbox.lock(TaskLockType.EXCLUSIVE, lowPriorityTask, interval);
+    Assert.assertFalse(lockResult.isOk());
```

Comment:
@jihoonson I see that this is expected behavior, but is this really ok. So, low priority task can't get the lock even after high priority task has unlocked.
I think, this behavior is creating following problem.

- a kill task was created that acquired the lock.
- a index task came and took same lock , revoking above before kill task even had a chance to run.
- overlord restarted for whatever reason
- kill task would not be given to taskRunner as it can't get lock , index task has it.
- now index task is finished , and unlocked.
- kill task is forever in waiting state because it is not being able to acquire the lock even if no-one has that lock.

Reply:
Hmm, so the intention of the revoked lock is for tasks to know that the acquired lock is no longer valid so the task should fail itself. I guess the real issue might be a wrong task state after the overlord is restarted. Once a kill task gets a lock, its isReady() method should be called already, which means, the kill task is at least in the pending state. i think this state shouldn’t be changed after restart.

Reply:
maybe, but given that the high priority task (which caused the revocation) is done, unlocked and gone.. why the low priority task shouldn't be able to get the lock now and do its thing instead of dying?

Reply:
I think current behavior makes sense for case ...
1. low priority indexing task1 took lock , started running creating segments.
2. high priority indexing task2 took lock, revoked above, created segments, published them
3. at this point segments created by indexing task1 might be outdated and hence it should not be able to make progress.

Reply:
however, this causes potential for task to be stuck in waiting state forever in case of overlord restarts... as task.isReady() will keep on return false forever due to the revoked lock.

Reply:
I guess there could be a couple of reasons. 

1) I guess we don't know exactly what users want. In this example, you might want to kill some segments after the index task is finished, but in the same scenario, some people want to just ignore the kill task and run the index task. Running kill task could be surprising for these people.
2) It's not easy to make the behavior consistent across the tasks of different states. For example, if some locks of a running task are revoked, to make the behavior consistent, the task could be paused for a while until the high priority task releases the lock and continues. I think this could be doable, but will need a lot of work.

Reply:
Oh I missed your last 2 comments. It makes sense too. Maybe we need to figure out why its state has changed after restart and fix it.

Reply:
I "think" here is what led to the problem state...

1. a kill task is submitted, TaskQueue called `isReady()` on it which created a lock in DB for it.
2. above task hasn't started running yet, a index task is submitted over same interval. TaskQueue called `isReady()` on it which revoked kill task's lock.
3. none of the above tasks are running just yet
4. overlord restarted
5. kill task's `isReady()` now returns false forever and `TaskQueue` never gives it to `TaskRunner`
5. index task's `isReady()` said true, it was run and finished successfully.

I think the problem happens because `isReady()` doesn't differentiate between not getting a lock and having a revoked lock . Ideally `TaskQueue` should notice that kill task's lock is revoked and should set its state to failed.

Reply:
Oh, I see. Perhaps this could be related to the lame task state tracking in the overlord. :sob:
I'm not sure if it's the best idea but I think we can change the return type of `isReady()` to some enum value which could be one of `READY`, `NOT_READY`, and `CANNOT_READY`.

Reply:
sorry , got sidetracked in some other stuff.

I can't think of a better quick fix so that would be OK I think.